pkg/reconciler/cache/replication/replication_reconcile.go

```diff
@@ -0,0 +1,133 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	kcpcache "github.com/kcp-dev/apimachinery/pkg/cache"
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+func (c *controller) reconcile(ctx context.Context, grKey string) error {
+	keyParts := strings.Split(grKey, "::")
+	if len(keyParts) != 2 {
+		return fmt.Errorf("incorrect key: %v, expected group.resource::key", grKey)
+	}
+	switch keyParts[0] {
+	case apisv1alpha1.SchemeGroupVersion.WithResource("apiexports").String():
+		return c.reconcileAPIExports(ctx, keyParts[1], apisv1alpha1.SchemeGroupVersion.WithResource("apiexports"))
+	default:
+		return fmt.Errorf("unsupported resource %v", keyParts[0])
+	}
+}
+
+// reconcileAPIExports makes sure that the ApiExport under the given key from the local shard is replicated to the cache server.
+// the replication function handles the following cases:
+//  1. creation of the object in the cache server when the cached object is not found in c.localApiExportLister
+//  2. deletion of the object from the cache server when the original/local object was removed OR was not found in c.localApiExportLister
+//  3. modification of the cached object to match the original one when meta.annotations, meta.labels, spec or status are different
+func (c *controller) reconcileAPIExports(ctx context.Context, key string, gvr schema.GroupVersionResource) error {
+	var cacheApiExport *apisv1alpha1.APIExport
+	var localApiExport *apisv1alpha1.APIExport
+	var cluster logicalcluster.Name
+	var err error
+
+	localApiExport, err = c.localApiExportLister.Get(key)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if errors.IsNotFound(err) {
+		var apiExportName string
+		cluster, _, apiExportName, err = kcpcache.SplitMetaClusterNamespaceKey(key)
+		if err != nil {
+			return err
+		}
+		cacheApiExport, err = c.retrieveApiExport(apiExportName, cluster.String())
+		if err != nil {
+			return err
+		}
+		// issue a live GET to make sure the localApiExport was removed
+		unstructuredLiveLocalApiExport, err := c.dynamicLocalClient.Cluster(cluster).Resource(gvr).Get(ctx, apiExportName, metav1.GetOptions{})
+		if err == nil {
+			return fmt.Errorf("the informer used by this controller is stale, the following APIExport was found on the local server: %s/%s/%s but was missing in the informer", cluster, unstructuredLiveLocalApiExport.GetNamespace(), unstructuredLiveLocalApiExport.GetName())
+		}
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	if cacheApiExport == nil {
+		cacheApiExport, err = c.retrieveApiExport(localApiExport.Name, logicalcluster.From(localApiExport).String())
```

Comment:
why don't we do this as first step? Then we don't need two calls here.

Reply:
done